### PR TITLE
[IOTDB-2680]Fix too many open files exception

### PR DESCRIPTION
**Pull request by choubenson**
Compaction breaks at 0.12.5, then recover at 0.13, it will report "too many open files" exception.

Solutions:
1. Limit num of unseq files in cross space compaction.
2. Manage all tsfile readers by FileReaderManager.

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/selector/RewriteCompactionFileSelector.java [hunk 1]
```diff
@@ -195,7 +198,9 @@ void select(boolean useTightBound) throws IOException {
   }
 
   private boolean updateSelectedFiles(long newCost, TsFileResource unseqFile) {
-    if (totalCost + newCost < memoryBudget) {
+    if (selectedSeqFiles.size() + selectedUnseqFiles.size() + 1 + tmpSelectedSeqFiles.size()
+            < maxCrossCompactionFileNum
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/selector/RewriteCompactionFileSelector.java:202]:
```suggestion
            <= maxCrossCompactionFileNum
```

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/utils/MultiTsFileDeviceIterator.java [hunk 2]
```diff
@@ -72,6 +74,23 @@ public MultiTsFileDeviceIterator(List<TsFileResource> tsFileResources) throws IO
...
+          FileReaderManager.getInstance().get(tsFileResource.getTsFilePath(), true);
+      readerMap.put(tsFileResource, reader);
+      deviceIteratorMap.put(tsFileResource, reader.getAllDevicesIteratorWithIsAligned());
+    }
+    for (TsFileResource tsFileResource : unseqResources) {
+      TsFileSequenceReader reader =
+          FileReaderManager.getInstance().get(tsFileResource.getTsFilePath(), true);
+      readerMap.put(tsFileResource, reader);
+      deviceIteratorMap.put(tsFileResource, reader.getAllDevicesIteratorWithIsAligned());
+    }
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/utils/MultiTsFileDeviceIterator.java:91]: catch throwable and close all readers

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/utils/MultiTsFileDeviceIterator.java [hunk 3]
```diff
@@ -72,6 +74,23 @@ public MultiTsFileDeviceIterator(List<TsFileResource> tsFileResources) throws IO
     }
   }
 
+  /** Used for cross space compaction. */
+  public MultiTsFileDeviceIterator(
```
Comment (THUMarkLau) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/utils/MultiTsFileDeviceIterator.java:78]: If exception occurs here, end query may close some files which have not yet build reader.

File: server/src/test/java/org/apache/iotdb/db/engine/compaction/utils/CompactionConfigRestorer.java [hunk 4]
```diff
@@ -58,7 +58,7 @@ public void restoreCompactionConfig() {
     config.setTargetChunkPointNum(targetChunkPointNum);
     config.setChunkSizeLowerBoundInCompaction(chunkSizeLowerBoundInCompaction);
     config.setChunkPointNumLowerBoundInCompaction(chunkPointNumLowerBoundInCompaction);
-    config.setMaxCompactionCandidateFileNum(maxCompactionCandidateFileNum);
+    config.setMaxInnerCompactionCandidateFileNum(maxCompactionCandidateFileNum);
     config.setConcurrentCompactionThread(concurrentCompactionThread);
```
Comment (THUMarkLau) [inline server/src/test/java/org/apache/iotdb/db/engine/compaction/utils/CompactionConfigRestorer.java:62]: Add `config.setMaxCrossCompactionCandidateFileNum(1000)` here

**Review by THUMarkLau — CHANGES_REQUESTED** [review]
change according to comment

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/selector/RewriteCompactionFileSelector.java (hunk 1 above)
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/selector/RewriteCompactionFileSelector.java:202]: Resolved

File: server/src/test/java/org/apache/iotdb/db/engine/compaction/utils/CompactionConfigRestorer.java (hunk 4 above)
Comment (choubenson) [reply server/src/test/java/org/apache/iotdb/db/engine/compaction/utils/CompactionConfigRestorer.java:62]: Resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/utils/MultiTsFileDeviceIterator.java (hunk 2 above)
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/utils/MultiTsFileDeviceIterator.java:91]: If the reader is closed in this method, there will be problems. Query f1, merge f1 and f2, and throw an exception when creating f2 reader in compaction. If f1 is closed in this method, the query will report an error. And f2 reader will be closed in end query.

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/utils/MultiTsFileDeviceIterator.java (hunk 3 above)
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/utils/MultiTsFileDeviceIterator.java:78]: If the reader is closed in this method, there will be problems. Query f1, merge f1 and f2, and throw an exception when creating f2 reader in compaction. If f1 is closed in this method, the query will report an error. And f2 reader will be closed in end query.

**Review by qiaojialin — APPROVED** [review]

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/utils/MultiTsFileDeviceIterator.java (hunk 3 above)
Comment (THUMarkLau) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/inner/utils/MultiTsFileDeviceIterator.java:78]: Get it

**Review by THUMarkLau — APPROVED** [review]

**Pull request merged**